src/main/scala/mesosphere/marathon/scheduling/Scheduler.scala

```diff
@@ -0,0 +1,67 @@
+package mesosphere.marathon
+package scheduling
+
+import akka.Done
+import mesosphere.marathon.core.instance.Instance
+import mesosphere.marathon.core.launcher.OfferProcessor
+import mesosphere.marathon.state.PathId
+import org.apache.mesos.Protos
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait Scheduler extends OfferProcessor {
+
+  //def create(runSpec: RunSpec): Future[Instance.Id]
+
+  /**
+    * Retrieve all instances for a specific run spec.
+    *
+    * @param runSpecId The path id of the run spec.
+    * @param ec The execution context for the future.
+    * @return A future list of all instances of the run spec.
+    */
+  def getInstances(runSpecId: PathId)(implicit ec: ExecutionContext): Future[Seq[Instance]]
+
+  /**
+    * Run all instances with given ids.
+    *
+    * This method is idempotent.
+    *
+    * @param instanceIds
+    * @param ec
+    * @return
+    */
+  def run(instanceIds: Instance.Id*)(implicit ec: ExecutionContext): Future[Done]
```

Comment:
How does one restart an instance?

Also, do we get a `Done` when the instance goal is updated, or when the instance is actually running?

Reply:
> Also, do we get a Done when the instance goal is updated, or when the instance is actually running?

Only when the instance goal changed.

> How does one restart an instance?

That is an open question. The orchestrator will receive instance updates before the scheduler does. So it could do:

1. Call `scheduler.stop(instanceId)`
2. Receive `case InstanceUpdate(instanceId, state.Running) => scheduler.run(instanceId)`

Reply:
I agree that restarting is an Orchestrator concern

Reply:
I don't think instances should ever change their run spec version (which does not work with reservations for now). So yes, it's an orchestrator concern to stop an instance and start a new one with a new version.